source/mainnet/smart-contracts/tutorials/fungible-tokens/smart-contract.rst

```diff
@@ -0,0 +1,517 @@
+        // Get the sender of the transaction
+        let sender = ctx.sender();
+
+        ensure!(sender.matches_account(&owner), ContractError::Unauthorized);
+
+        // Parse the parameter.
+        let params: MintParams = ctx.parameter_cursor().get()?;
+
+        let (state, builder) = host.state_and_builder();
+        for (token_id, token_info) in params.tokens {
+            // ensure!(
+            //     state.contains_token(&token_id),
+            //     ContractError::Custom(CustomContractError::TokenAlreadyMinted)
+            // );
+
+            if !state.contains_token(&token_id) {
+                state.set_max_supply(&token_id, token_info.1.max_supply)
+            } else {
+                let max_supply = state.get_token_supply(&token_id)?;
+                let circulating_suppy = state.get_circulating_supply(&token_id)?;
+
+                ensure!(
+                    circulating_suppy <= max_supply,
+                    ContractError::Custom(CustomContractError::MaxSupplyReached)
+                );
+
+                ensure!(
+                    &token_info.1.amount <= &(max_supply - circulating_suppy),
+                    ContractError::Custom(CustomContractError::MaxSupplyReached)
+                );
+            }
+
+            // Mint the token in the state.
+            state.mint(
+                &token_id,
+                &token_info.0,
+                token_info.1.amount,
+                &params.owner,
+                builder,
+            );
+
+            // Event for minted token.
+            logger.log(&Cis2Event::Mint(MintEvent {
+                token_id,
+                amount: token_info.1.amount,
+                owner: params.owner,
+            }))?;
+
+            // Metadata URL for the token.
+            logger.log(&Cis2Event::TokenMetadata::<_, ContractTokenAmount>(
+                TokenMetadataEvent {
+                    token_id,
+                    metadata_url: token_info.0.to_metadata_url(),
+                },
+            ))?;
+        }
+        Ok(())
+    }
+
+In the following ``if`` clause you are checking if this token has been minted before. If not, set its maximum supply in the state by using the ``set_max_supply()`` function.
+
+.. code-block:: rust
+
+    if !state.contains_token(&token_id) {
+                state.set_max_supply(&token_id, token_info.1.max_supply)
+            }
+    else {
+                let max_supply = state.get_token_supply(&token_id)?;
+                let circulating_supply = state.get_circulating_supply(&token_id)?;
+
+                ensure!(
+                    circulating_supply <= max_supply,
+                    ContractError::Custom(CustomContractError::MaxSupplyReached)
+                );
+
+                ensure!(
+                    &token_info.1.amount <= &(max_supply - circulating_supply),
+                    ContractError::Custom(CustomContractError::MaxSupplyReached)
+                );
+            }
+
```

Comment:
The indentation here seems off and the code is not formatted as usual. Can we fix this?

Reply:
I've copied the code example again from Medium. If it is still incorrect, then it is also incorrect on Medium. I am not doing anything to the code once I copy it.

Reply:
Hey sorry for the inconvenience, can fix that on the medium and the code snippet can be found (formatted) in my profile, https://github.com/bogacyigitbasi/cis2-fungible/blob/main/src/lib.rs#L563

Reply:
I've replaced the example with the code on github above. @abizjak could you check to see if it appears as you expect now?

Reply:
I added a suggestion which I think looks better (without the extra spaces).